blocklytest/src/androidTest/java/com/google/blockly/model/BlockFactoryTest.java

```diff
@@ -65,154 +65,194 @@ public void testLoadBlocks() {
+            BlockTestStrings.VALUE_SHADOW_GOOD));
+        Connection conn = loaded.getInputByName("value_input").getConnection();
         assertEquals("VALUE_REAL", conn.getTargetBlock().getId());
         assertFalse(conn.getTargetBlock().isShadow());
         assertEquals("VALUE_SHADOW", conn.getShadowBlock().getId());
         assertTrue(conn.getShadowBlock().isShadow());
+    }
 
-        loaded = parseBlockFromXml(BlockTestStrings.assembleFrankenblock("block", "3",
-                BlockTestStrings.STATEMENT_SHADOW));
-        conn = loaded.getInputByName("NAME").getConnection();
+    public void testLoadFromXmlShadowInStatement() throws IOException, XmlPullParserException {
+        Block loaded = parseBlockFromXml(BlockTestStrings.assembleFrankenblock("block", "3",
+            BlockTestStrings.STATEMENT_SHADOW));
+        Connection conn = loaded.getInputByName("NAME").getConnection();
         assertEquals(conn.getTargetBlock(), conn.getShadowBlock());
         assertTrue(conn.getShadowBlock().isShadow());
+    }
 
-        // Clear refs so block names can be reused
-        mBlockFactory.clearPriorBlockReferences();
-        loaded = parseBlockFromXml(BlockTestStrings.assembleFrankenblock("block", "4",
-                BlockTestStrings.STATEMENT_SHADOW_GOOD));
-        conn = loaded.getInputByName("NAME").getConnection();
+    public void testLoadFromXmlShadowWithBlockInStatement() throws IOException, XmlPullParserException {
+        Block loaded = parseBlockFromXml(BlockTestStrings.assembleFrankenblock("block", "4",
+            BlockTestStrings.STATEMENT_SHADOW_GOOD));
+        Connection conn = loaded.getInputByName("NAME").getConnection();
         assertEquals("STATEMENT_REAL", conn.getTargetBlock().getId());
         assertFalse(conn.getTargetBlock().isShadow());
         assertEquals("STATEMENT_SHADOW", conn.getShadowBlock().getId());
         assertTrue(conn.getShadowBlock().isShadow());
+    }
 
-        loaded = parseBlockFromXml(BlockTestStrings.assembleFrankenblock("block", "5",
-                BlockTestStrings.VALUE_NESTED_SHADOW));
-        conn = loaded.getInputByName("value_input").getConnection();
+    public void testLoadFromXmlNestedShadow() throws IOException, XmlPullParserException {
+        Block loaded = parseBlockFromXml(BlockTestStrings.assembleFrankenblock("block", "5",
+            BlockTestStrings.VALUE_NESTED_SHADOW));
+        Connection conn = loaded.getInputByName("value_input").getConnection();
         assertEquals(conn.getTargetBlock(), conn.getShadowBlock());
         Block shadow1 = conn.getShadowBlock();
         assertEquals("SHADOW1", shadow1.getId());
         conn = shadow1.getOnlyValueInput().getConnection();
         assertEquals(conn.getTargetBlock(), conn.getShadowBlock());
         assertEquals("SHADOW2", conn.getShadowBlock().getId());
+    }
 
-        // Clear refs so block names can be reused
-        mBlockFactory.clearPriorBlockReferences();
```

Comment:
This didn't seem to be required anymore. Happy to add it back in if I've misunderstood.